Lib/asyncio/futures.py

```diff
@@ -51,6 +51,8 @@ class Future:
     _exception = None
     _loop = None
     _source_traceback = None
+    # A saved CancelledError for later chaining.
+    _cancelled_exc = None
```

Comment:
I'd make this a wekref to break potential garbage cycles. Sometimes tasks are cancelled in bulk but their cancelled errors aren't retrieved -- in this case we'd like them to GC asap.

Reply:
Thanks for the feedback here. I'm having trouble getting the desired traceback behavior when using a weakref here though.

When using a weakref, the traceback seems to work okay in a stand-alone script when the `_asynciomodule.c` accelerator module is completely disabled.

However, when running tests, I'm not able to get the new test cases to pass with any of the `PyTask_*` test classes. This is because for some reason, if a `Future` is cancelled, it seems like the exception object doesn't hang around long enough with weakref if it's a `_CFuture` (C implementation), but it does if it's a `_PyFuture` (pure Python implementation). This causes the weakly referenced exception not to be available when arriving at the code to set `exc.__context__`. Also, I found that this happens in _all_ of the `PyTask_*` versions (even `PyTask_PyFuture_Tests`) because even when running `PyFuture` tests, `asyncio.sleep()` (which is used widely in the tests) still uses `_CFuture` under the hood because `loop.create_future()` uses a hard-coded Future class. (I'm not sure if that was intended or not.)

Must weakref be figured out before the 3.9 beta deadline?